feat: auto-refresh on new revisions during sync retries (Alpha) (#11494)

aslafy-z opened the pull request:
This feature allows users to configure their apps to refresh on new revisions, when the current sync is retrying.
This is controlled by a new boolean Application CRD field `syncPolicy.retry.refresh` or via the `--sync-retry-refresh` flag.

Closes https://github.com/argoproj/argo-cd/issues/11494
Related to https://github.com/argoproj/argo-cd/issues/6055
Discussed at https://www.youtube.com/watch?v=baIX9Bk6f5w&t=1173s 

Initially based of the work Sayrus did at https://github.com/Sayrus/argo-cd/commit/817bc3449768021d0d5ad7f1ce7510bcd9d2f486

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

aslafy-z commented on controller/appcontroller.go:
```diff
@@ -1195,6 +1195,14 @@ func (ctrl *ApplicationController) processRequestedAppOperation(app *appv1.Appli
 				ctrl.requestAppRefresh(app.QualifiedName(), CompareWithLatest.Pointer(), &retryAfter)
 				return
 			} else {
+				if state.Phase == synccommon.OperationRunning && app.Spec.SyncPolicy.Retry.Refresh && app.Status.Sync.Revision != state.Operation.Sync.Revision {
```
This works for classic resources, not hooks. Are hooks processed somewhere else?

aslafy-z commented on controller/appcontroller.go:
```diff
@@ -1195,6 +1195,14 @@ func (ctrl *ApplicationController) processRequestedAppOperation(app *appv1.Appli
 			} else {
+				if state.Phase == synccommon.OperationRunning && app.Spec.SyncPolicy.Retry.Refresh && app.Status.Sync.Revision != state.Operation.Sync.Revision {
+					logCtx.Infof("A new revision is available, refreshing and terminating app, was phase: %s, message: %s", state.Phase, state.Message)
+					ctrl.requestAppRefresh(app.QualifiedName(), CompareWithLatest.Pointer(), nil)
```
This forces a refresh, without it, it waits for the next window, which can take some time.

aslafy-z commented on controller/appcontroller.go:
```diff
@@ -1195,6 +1195,14 @@ func (ctrl *ApplicationController) processRequestedAppOperation(app *appv1.Appli
+					logCtx.Infof("A new revision is available, refreshing and terminating app, was phase: %s, message: %s", state.Phase, state.Message)
+					ctrl.requestAppRefresh(app.QualifiedName(), CompareWithLatest.Pointer(), nil)
+					state.Phase = synccommon.OperationTerminating
+					state.Message = "Operation forced to terminate (new revision available)"
```
I could not find this message on the UI, is there a way to make it appear?

andrii-korotkov-verkada replied:
I think we have to move `setOperationState` above the `requestAppRefresh`, since otherwise there can be a race condition by the refresh processor kicking in and message not being ready yet.

aslafy-z commented:
@alexec any inputs?

erchn commented on assets/swagger.json:
```diff
@@ -8471,6 +8471,10 @@
+        },
+        "refresh": {
+          "type": "boolean",
+          "title": "Refresh indicates wether a new revision should trigger a new sync (default: false)"
```
```suggestion
          "title": "Refresh indicates whether a new revision should trigger a new sync (default: false)"
```

aslafy-z commented:
Hi there, Zadkiel here. I originally signed up to talk about this topic in the contributor meeting but finally could not make it.
Here is the recording: https://youtu.be/tJYxtn3cO6E?t=1087
This is a very annoying missing feature on the ArgoCD side that in my opinion breaks the experience of auto-sync.

Please give feedback so we can iterate on this topic! 🙏

pasha-codefresh commented:
Hey @aslafy-z , thank you for this PR, really amazing feature. 
I have tried to test it , and it not really works for me, 

I have guestbook application with simple PreSync hook 

`apiVersion: batch/v1
kind: Job
metadata:
  name: before
  annotations:
    argocd.argoproj.io/hook: PreSync
    argocd.argoproj.io/hook-delete-policy: BeforeHookCreation
spec:
  template:
    spec:
      containers:
        - name: sleep
          image: alpine:latest
          command: ["sleep", "30", "exit", "1"]
      restartPolicy: Never
  backoffLimit: 0
`

And once i am doing new commit, i am almost immediately getting such error

<img width="1279" alt="Снимок экрана 2024-04-29 в 17 55 52" src="https://github.com/argoproj/argo-cd/assets/39732895/86cad8a6-614e-43d0-85ae-d1dae7688710">

aslafy-z commented:
@pasha-codefresh Thank you for giving attention to this PR. I reported in the PR comments that it was not triggering refresh when processing hooks: https://github.com/argoproj/argo-cd/pull/15603#discussion_r1340694530
I'm not sure how to resolve it, and would like some experimented contributor or maintainer to help here.

akorzy-pl commented:
Looking forward to seeing this merged, even if it doesn't cover all the use cases :-)

pasha-codefresh commented:
Sorry @aslafy-z , missed it!

pasha-codefresh commented:
@aslafy-z PR looks good to me, could you please add documentation about this feature?

https://argo-cd.readthedocs.io/en/stable/user-guide/auto_sync/

todaywasawesome commented on manifests/core-install.yaml:
```diff
@@ -107,6 +107,10 @@ spec:
                     format: int64
                     type: integer
+                  refresh:
+                    description: 'Refresh indicates wether a new revision should trigger
```
```suggestion
                    description: 'Refresh indicates if a new revision should trigger
```

aslafy-z replied:
I applied the change!

todaywasawesome commented on manifests/core-install.yaml:
```diff
@@ -1668,6 +1672,10 @@ spec:
                         format: int64
                         type: integer
+                      refresh:
+                        description: 'Refresh indicates wether a new revision should
```
```suggestion
                        description: 'Refresh indicates if a new revision should
```

todaywasawesome commented on manifests/core-install.yaml:
```diff
@@ -2529,6 +2537,10 @@ spec:
                             format: int64
                             type: integer
+                          refresh:
+                            description: 'Refresh indicates wether a new revision
```
```suggestion
                            description: 'Refresh indicates if a new revision
```

todaywasawesome commented on manifests/crds/application-crd.yaml:
```diff
@@ -106,6 +106,10 @@ spec:
                     format: int64
                     type: integer
+                  refresh:
+                    description: 'Refresh indicates wether a new revision should trigger
```
```suggestion
                    description: 'Refresh indicates if a new revision should trigger
```

todaywasawesome commented on manifests/crds/application-crd.yaml:
```diff
@@ -1667,6 +1671,10 @@ spec:
                         format: int64
                         type: integer
+                      refresh:
+                        description: 'Refresh indicates wether a new revision should
```
```suggestion
                        description: 'Refresh indicates if a new revision should
```

todaywasawesome commented on manifests/crds/application-crd.yaml:
```diff
@@ -2528,6 +2536,10 @@ spec:
                             format: int64
                             type: integer
+                          refresh:
+                            description: 'Refresh indicates wether a new revision
```
```suggestion
                            description: 'Refresh indicates if a new revision
```

todaywasawesome commented on manifests/ha/install.yaml:
```diff
@@ -107,6 +107,10 @@ spec:
                     format: int64
                     type: integer
+                  refresh:
+                    description: 'Refresh indicates wether a new revision should trigger
```
```suggestion
                    description: 'Refresh indicates if a new revision should trigger
```

todaywasawesome commented on manifests/ha/install.yaml:
```diff
@@ -1668,6 +1672,10 @@ spec:
                         format: int64
                         type: integer
+                      refresh:
+                        description: 'Refresh indicates wether a new revision should
```
```suggestion
                        description: 'Refresh indicates if a new revision should
```

todaywasawesome commented on manifests/ha/install.yaml:
```diff
@@ -2529,6 +2537,10 @@ spec:
                             format: int64
                             type: integer
+                          refresh:
+                            description: 'Refresh indicates wether a new revision
```
```suggestion
                            description: 'Refresh indicates if a new revision
```

todaywasawesome commented on manifests/install.yaml:
```diff
@@ -107,6 +107,10 @@ spec:
                     format: int64
                     type: integer
+                  refresh:
+                    description: 'Refresh indicates wether a new revision should trigger
```
```suggestion
                    description: 'Refresh indicates if a new revision should trigger
```

todaywasawesome commented on manifests/install.yaml:
```diff
@@ -1668,6 +1672,10 @@ spec:
                         format: int64
                         type: integer
+                      refresh:
+                        description: 'Refresh indicates wether a new revision should
```
```suggestion
                        description: 'Refresh indicates if a new revision should
```

todaywasawesome commented on manifests/install.yaml:
```diff
@@ -2529,6 +2537,10 @@ spec:
                             format: int64
                             type: integer
+                          refresh:
+                            description: 'Refresh indicates wether a new revision
```
```suggestion
                            description: 'Refresh indicates if a new revision
```

todaywasawesome reviewed (changes requested):
I think "whether" is the wrong word to use and it is misspelled. If you're ok with the change you can just accept the commit.

aslafy-z commented:
@todaywasawesome I applied the required change.
@pasha-codefresh I added a sparse documentation for the feature, let me know if it's OK to you.

I also added the `argocd app set --sync-retry-refresh` flag.

aslafy-z commented:
I will fix the tests asap

pasha-codefresh commented:
@aslafy-z any news?

pasha-codefresh commented:
hey @aslafy-z , are you going to work on it?

aslafy-z commented:
Hi @pasha-codefresh,

I couldn't reproduce the issue correctly in the end-to-end tests. I'm quite short on time these days. Please feel free to take over if needed, and any suggestions are welcome.

oleghalin commented:
> Hi @pasha-codefresh,
> 
> I couldn't reproduce the issue correctly in the end-to-end tests. I'm quite short on time these days. Please feel free to take over if needed, and any suggestions are welcome.

I think your test case shouldnt wait for `Failed` state because patched selector will not be applied and you will stuck at `Running` state.

```
error when patching "/dev/shm/919124178": Deployment.apps "guestbook-ui" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app":"guestbook-ui", "foo":"bar"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable
```

oleghalin commented on test/e2e/app_autosync_test.go:
```diff
@@ -92,5 +92,49 @@ func TestAutoSyncSelfHealEnabled(t *testing.T) {
+		Expect(SyncStatusIs(SyncStatusCodeSynced)).
+		When().
+		// app should be attempted to auto-synced once and marked with error after failed attempt detected
+		PatchFile("guestbook-ui-deployment.yaml", `[{"op": "add", "path": "/spec/selector/matchLabels/foo", "value": "bar"}, {"op": "add", "path": "/spec/template/metadata/labels/foo", "value": "bar"}]`).
+		Refresh(RefreshTypeNormal).
+		Then().
+		Expect(OperationPhaseIs(OperationFailed)).
+		When().
+		// Trigger refresh again to make sure controller notices previously failed sync attempt before expectation timeout expires
+		Refresh(RefreshTypeNormal).
+		Then().
+		Expect(SyncStatusIs(SyncStatusCodeOutOfSync)).
+		Expect(Condition(ApplicationConditionSyncError, "Failed sync attempt")).
```
```suggestion
		PatchFile("guestbook-ui-deployment.yaml", `[{"op": "add", "path": "/spec/selector/matchLabels/foo", "value": "bar"}, {"op": "add", "path": "/spec/template/metadata/labels/foo", "value": "bar"}]`).
		Refresh(RefreshTypeNormal).
		Then().
		Expect(OperationPhaseIs(OperationRunning)).
		When().
```

I suppose that you copied this part from previous test and your case doesnt require it. 

- Applying healthy manifest
- Waiting for Healthy state
- Patching with corrupted manifest (this case application will stuck in `Syncing` state. (operationState.phase == 'Running')
- Expecting phase running
- Patching with healthy manifest
- Expecting that introduced feature works as expected (previous sync stopped and new started & application goes to healthy) 


Let me know if it makes sense

andrii-korotkov-verkada commented on controller/appcontroller.go:
```diff
@@ -1294,6 +1294,15 @@ func (ctrl *ApplicationController) processRequestedAppOperation(app *appv1.Appli
+					state.Message = "Operation forced to terminate (new revision available)"
+					ctrl.setOperationState(app, state)
+					return
+				}
```
There's a setOperationState below as well, let's combine the two.

olivergondza commented:
@aslafy-z, @giepa, Thanks a lot for pushing the RFE this far.

Do you have plans to follow up with this? Any work in progress you can share on top of this PR? Thanks!

chaehni commented:
I don't think tying this to the retry logic is a good thng. The problem is orthogonal to retries. For more context, see: https://github.com/argoproj/argo-cd/issues/11494#issuecomment-2813729843

olivergondza commented:
Can be closed in favor of https://github.com/argoproj/argo-cd/pull/23038

aslafy-z commented:
Thank you for taking over @olivergondza!
